### DISCUSSION TITLE
'Unmapped target property' when using CollectionMappingStrategy.ADDER_PREFERRED

### DISCUSSION THREAD

**Discussion by JoshukLite** [mapstruct/mapstruct#3699] [Q&A]
Hi people,
I'm facing a small issue when using `collectionMappingStrategy = CollectionMappingStrategy.ADDER_PREFERRED`.
Language - Java, version 21.
Mapstruct version - 1.5.2.
I'm using it to map objects from DTO to JPA Data. My model has a collection - the PersonData class has multiple AddressData objects (one to many relations). As I'm using JPA, I need to use a helper method to correctly add an AddressData object to the PersonData addresses list, so AddressData will have a link to its PersonData object.
Here is what my code looks like:
```
@Mapper(collectionMappingStrategy = CollectionMappingStrategy.ADDER_PREFERRED)
public interface PersonPersistenceMapper {
    PersonData to(PersonDto personDto);
}
```
```
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class PersonData {

    List<AddressData> addresses;

    public void addAddress(AddressData address) {
        addresses.add(address);
        address.setPerson(this);
    }
}
```
```
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class AddressData {

    private String street;

    private PersonData person;

}
```
```
@Data
public class PersonDto {
    private List<AddressDto> addresses;
}
```
```
@Data
public class AddressDto {
    private String street;
}
```

Mapstruct creates the perfect implementation:
```
public class PersonPersistenceMapperImpl implements PersonPersistenceMapper {

    @Override
    public PersonData to(PersonDto personDto) {
        if ( personDto == null ) {
            return null;
        }

        PersonData personData = new PersonData();

        if ( personDto.getAddresses() != null ) {
            for ( AddressDto address : personDto.getAddresses() ) {
                personData.addAddress( addressDtoToAddressData( address ) );
            }
        }

        return personData;
    }

    protected AddressData addressDtoToAddressData(AddressDto addressDto) {
        if ( addressDto == null ) {
            return null;
        }

        AddressData addressData = new AddressData();

        addressData.setStreet( addressDto.getStreet() );

        return addressData;
    }
}
```
While it generates mapping methods for the collection field, it shows unmapped warning though
```
warning: Unmapped target property: "street". Mapping from property "List<AddressDto> addresses" to "AddressData addresses".
    PersonData to(PersonDto personDto);
```

**Comment by filiphr** [comment]
Thanks for the detailed example @JoshukLite. That looks strange and indeed it is incorrect. Can you please try this out with 1.6.0?

**Comment by JoshukLite** [comment]
@filiphr 
It was my mistake. The real warning was that `AddressData.person` field was not explicitly set, and indeed it wasn't as it was implicitly done within `AdressData.addAddress()` method.
The only warning I have right now is:
```
Unmapped target property: "person". Mapping from property "List<AddressDto> addresses" to "AddressData addresses".
    PersonData to(PersonDto personDto);
```
This is correct, as it's not MapStruct who sets it, it's the code of the `AdressData.addAddress()` method, thus I need explicitly ignore it using MapStruct `@Mapping` annotation.

Thanks for the help. This thread can be closed now.
